### ISSUE TITLE
Interrupting a workflow instance waiting on a parallel (joining) gateway leads to a stuck instance

### ISSUE THREAD

**Issue by saig0** [zeebe-io/zeebe#4400]
**Describe the bug**
When the workflow instance is interrupted while waiting on a parallel (joining) gateway then the workflow instance can not be completed/terminated.

![image](https://user-images.githubusercontent.com/4305769/80495158-d9fb8e00-8967-11ea-88d0-5d74a2c542d3.png)

In this case, the taken incoming sequence flow of the gateway is taken again after the sub-process was terminated.

![image](https://user-images.githubusercontent.com/4305769/80496470-8a1dc680-8969-11ea-9a17-d5e21167b456.png)

**To Reproduce**
* deploy a workflow with a parallel (joining) gateway inside a subprocess that has an interrupting boundary event
* create an instance
* process the workflow until one of the incoming sequence flows of the gateway is taken
* trigger the boundary event
* verify that all elements are completed/terminated except the workflow instance itself
* verify that the sequence flow is taken twice

**Expected behavior**
The workflow instance can be completed/terminated. 
The sequence flows are taken only once.

**Log/Stacktrace**

<details><summary>JUnit Engine Test</summary>
 <p>

```
@Test
  public void test() {
    ENGINE
        .deployment()
        .withXmlResource(
            Bpmn.readModelFromStream(getClass().getResourceAsStream("/workflows/Test_2.bpmn")))
        .deploy();

    final var wfInstance =
        ENGINE
            .workflowInstance()
            .ofBpmnProcessId("Process_0jszb0c-v2")
            .withVariable("inspections_id", "id-123")
            .create();

    ENGINE.job().ofInstance(wfInstance).withType("prepare_inspections").complete();

    ENGINE.job().ofInstance(wfInstance).withType("weight").complete();

    RecordingExporter.workflowInstanceRecords()
        .withWorkflowInstanceKey(wfInstance)
        .withElementId("SequenceFlow_1d2vkl1")
        .withIntent(WorkflowInstanceIntent.SEQUENCE_FLOW_TAKEN)
        .await();

    ENGINE.message().withName("timeout").withCorrelationKey("id-123").publish();

    RecordingExporter.workflowInstanceRecords()
        .withWorkflowInstanceKey(wfInstance)
        .filterRootScope()
        .withIntent(WorkflowInstanceIntent.ELEMENT_COMPLETED)
        .await();
  }
```

</p>
</details>

[Test_2.bpmn.txt](https://github.com/zeebe-io/zeebe/files/4546187/Test_2.bpmn.txt)

**Environment:**
 - OS: [e.g. Linux]
 - Zeebe Version: 0.22.1, 0.23.1
 - Configuration: [e.g. exporters etc.]

**Comment by saig0** [comment]
This was reported in the forum: https://forum.zeebe.io/t/boundary-event-doesnt-cancel-sub-process/1174

**Comment by saig0** [comment]
## Hints

* when a sequence flow on a parallel joining gateway is taken then it defers a record in the state to remember that this sequence flow was taken
* when a sub-process is terminated then it publishes all deferred records (under the assumption that only a triggered boundary event was deferred)
* when a deferred record is published then a new token is spawned
* a scope or the workflow instance can only be completed/terminated if all tokens are consumed

**Comment by saig0** [comment]
Related to #4099 4352

**Comment by saig0** [comment]
The cause of the problem is the same as #4352. On terminating the sub-process, a new token is spawned for the sequence flow that is waiting on the joining parallel gateway. As result, the flow scope of the sub-process can not be completed or terminated.
